### ISSUE TITLE
Cyclic reference with loop, resolver will never be destructed

### ISSUE THREAD

**Issue by brstgt** [amphp/dns#68]
As long as the gcWatcher is not cancelled, the BasicResolver will never be destroyed as the watcher contains a reference to the resolver which is retained by the loop. The loop driver will also never be destroyed as long it has watchers. Watcher callbacks have references to the loop and the loop has references to the watchers (or events). But this is something different.

**Comment by kelunik** [comment]
This is a valid concern, as a DNS resolver might be used standalone without relying on the global instance.

**Comment by brstgt** [comment]
Also it is not possible to correctly clean loop/resolver instances e.g. when it is replaced with another instance which happens massively in tests.

**Comment by brstgt** [comment]
Tested and works for manual cleanup:

```
    public function shutdown() {
        if ($this->gcWatcher !== null) {
            Loop::cancel($this->gcWatcher);
            $this->gcWatcher = null;
        }
    }

    public function __destruct() {
        $this->shutdown();
    }

```

**Comment by kelunik** [comment]
@brstgt Cyclic references shouldn't matter in tests and should be cleaned up correctly during the loop switch.

**Comment by kelunik** [comment]
I've just released v0.9.10 with the fix.
